pkg/graveler/ref/tag_iterator.go

```diff
@@ -2,118 +2,82 @@ package ref
 
 import (
 	"context"
-	"errors"
 
-	"github.com/treeverse/lakefs/pkg/db"
 	"github.com/treeverse/lakefs/pkg/graveler"
+	"github.com/treeverse/lakefs/pkg/kv"
 )
 
-type TagIterator struct {
-	db           db.Database
-	ctx          context.Context
-	repositoryID graveler.RepositoryID
-	value        *graveler.TagRecord
-	buf          []*graveler.TagRecord
-	offset       string
-	fetchSize    int
-	err          error
-	state        iteratorState
+type KVTagIterator struct {
+	it     kv.MessageIterator
+	err    error
+	value  *graveler.TagRecord
+	repoID graveler.RepositoryID
+	store  kv.Store
 }
 
-type tagRecord struct {
-	TagID    graveler.TagID    `db:"id"`
-	CommitID graveler.CommitID `db:"commit_id"`
-}
-
-func NewTagIterator(ctx context.Context, db db.Database, repositoryID graveler.RepositoryID, fetchSize int) *TagIterator {
-	return &TagIterator{
-		db:           db,
-		ctx:          ctx,
-		repositoryID: repositoryID,
-		fetchSize:    fetchSize,
-		buf:          make([]*graveler.TagRecord, 0, fetchSize),
+func NewKVTagIterator(ctx context.Context, store *kv.StoreMessage, repositoryID graveler.RepositoryID) (*KVTagIterator, error) {
+	it, err := kv.NewPrimaryIterator(ctx, store.Store, (&graveler.TagData{}).ProtoReflect().Type(),
+		graveler.TagPartition(repositoryID),
+		graveler.TagPath(""), "", true)
+	if err != nil {
+		return nil, err
 	}
+	return &KVTagIterator{
+		it:     it,
+		store:  store.Store,
+		repoID: repositoryID,
+	}, nil
 }
 
-func (ri *TagIterator) Next() bool {
-	if ri.err != nil {
+func (i *KVTagIterator) Next() bool {
+	if i.Err() != nil {
 		return false
 	}
-
-	ri.maybeFetch()
-
-	// stage a value and increment offset
-	if len(ri.buf) == 0 {
+	if !i.it.Next() {
+		i.value = nil
 		return false
 	}
-	ri.value = ri.buf[0]
-	ri.buf = ri.buf[1:]
-	ri.offset = string(ri.value.TagID)
-	return true
-}
-
-func (ri *TagIterator) maybeFetch() {
-	if ri.state == iteratorStateDone {
-		return
-	}
-	if len(ri.buf) > 0 {
-		return
-	}
-
-	var offsetCondition string
-	if ri.state == iteratorStateInit {
-		offsetCondition = iteratorOffsetCondition(true)
-		ri.state = iteratorStateQuerying
-	} else {
-		offsetCondition = iteratorOffsetCondition(false)
-	}
-
-	var buf []*tagRecord
-	err := ri.db.Select(ri.ctx, &buf, `
-			SELECT id, commit_id
-			FROM graveler_tags
-			WHERE repository_id = $1
-			AND id `+offsetCondition+` $2
-			ORDER BY id ASC
-			LIMIT $3`, ri.repositoryID, ri.offset, ri.fetchSize)
-	if err != nil {
-		ri.err = err
-		return
+	e := i.it.Entry()
+	if e == nil {
+		i.err = graveler.ErrNotFound
```

Comment:
Why `graveler.ErrNotFound` relevant here? I don't get it.

Reply:
added a new error instead